File: netcafe.py
"""
netcafe.py — 给 AI agent 用的一套工具,单文件、零依赖(只用标准库)。

不需要注册、不需要密钥就能开始用:匿名有免费额度,每次调用都会告诉你花了多少钱。
额度不够时,拿一把 AllRouter Key 传进来即可,调用方式完全不变。

    from netcafe import NetCafe
    nc = NetCafe()                      # 匿名,免费额度
    nc = NetCafe(key="sk-...")          # 带 Key,额度更大 + 跨设备记忆

    nc.search("MCP 协议 2026")           # 联网搜索
    nc.read("https://example.com")       # 网页转干净 Markdown
    nc.diagram("graph TD; A-->B")        # 图表 → 图片 URL(可直接嵌 Markdown)
    nc.transcribe("https://x.com/a.mp3") # 语音转文字
    nc.grammar("He go to school")        # 语法检查
    nc.pdf(url="https://example.com")    # 网页转 PDF
    nc.ask("一句话解释 MCP")              # 单模型问答,返回答案+精确花费
    nc.compare("同一个问题", models=[...]) # 多模型对比:答案/成本/延迟
    nc.remember("我偏好用便宜模型跑批量")   # 跨会话记忆
    nc.recall("我偏好什么模型")
    nc.build("一个番茄钟,可自定义时长")     # ★ 一句话造一个真正在线的应用

主页与全部工具文档:https://ainetcafe.com/mcp.html
"""

import json
import time
import urllib.parse
import urllib.request

__version__ = "1.0.0"
BASE = "https://ainetcafe.com"


class NetCafeError(RuntimeError):
    """调用失败。quota_exhausted 时 .how_to_continue 里有继续使用的办法。"""

    def __init__(self, message, payload=None):
        super().__init__(message)
        self.payload = payload or {}
        self.how_to_continue = self.payload.get("how_to_continue")
        self.tell_your_human = self.payload.get("tell_your_human")


class NetCafe:
    def __init__(self, key=None, base=BASE, timeout=120, source="pysdk"):
        self.key = key
        self.base = base.rstrip("/")
        self.timeout = timeout
        self.source = source

    # —— 内部 ——
    def _headers(self):
        h = {"user-agent": f"netcafe-python/{__version__}", "accept": "application/json"}
        if self.key:
            h["authorization"] = f"Bearer {self.key}"
        return h

    def _get(self, tool, **params):
        params = {k: v for k, v in params.items() if v is not None}
        params.setdefault("s", self.source)
        url = f"{self.base}/t/{tool}?{urllib.parse.urlencode(params)}"
        req = urllib.request.Request(url, headers=self._headers())
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as r:
                return json.loads(r.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            body = e.read().decode("utf-8", "replace")
            try:
                payload = json.loads(body)
            except ValueError:
                payload = {"error": body[:300]}
            raise NetCafeError(payload.get("error", f"HTTP {e.code}"), payload) from None

    def _rpc(self, tool, args):
        body = json.dumps({"jsonrpc": "2.0", "id": 1, "method": "tools/call",
                           "params": {"name": tool, "arguments": args}}).encode()
        headers = dict(self._headers())
        headers["content-type"] = "application/json"
        headers["accept"] = "application/json, text/event-stream"
        req = urllib.request.Request(f"{self.base}/mcp", data=body, headers=headers)
        with urllib.request.urlopen(req, timeout=self.timeout) as r:
            d = json.loads(r.read().decode("utf-8"))
        result = d.get("result", {})
        payload = result.get("structuredContent")
        if payload is None:
            content = (result.get("content") or [{}])[0].get("text", "{}")
            try:
                payload = json.loads(content)
            except ValueError:
                payload = {"text": content}
        if payload.get("error"):
            raise NetCafeError(payload["error"], payload)
        return payload

    # —— 读取类:一个 GET 就够 ——
    def search(self, query, max_results=8):
        """联网搜索,返回 [{title, url, snippet}]。"""
        return self._get("web_search", query=query, max_results=max_results).get("results", [])

    def read(self, url):
        """任意公开网页 → 干净 Markdown。"""
        return self._get("fetch_page", url=url).get("markdown", "")

    def diagram(self, source, type="mermaid", format="svg"):
        """图表代码 → 图片 URL,可直接写进 Markdown 的 ![](...)。"""
        return self._get("render_diagram", source=source, type=type, format=format).get("image_url")

    def diagram_url(self, source, type="mermaid", format="svg"):
        """不发请求,直接拼出无状态图片 URL(适合写死在 README 里)。"""
        import base64
        b64 = base64.urlsafe_b64encode(source.encode()).decode().rstrip("=")
        return f"{self.base}/i/{type}/{b64}?format={format}"

    def grammar(self, text, language="auto"):
        """语法/拼写/风格检查,返回逐条问题与修改建议。"""
        return self._get("check_grammar", text=text, language=language).get("issues", [])

    def transcribe(self, url, language=None):
        """音频 URL → 文字稿(自托管 Whisper)。"""
        return self._get("transcribe_audio", url=url, language=language).get("text", "")

    def pdf(self, url=None, html=None):
        """网页或 HTML → PDF,返回下载链接。"""
        return self._get("convert_to_pdf", url=url, html=html).get("pdf_url")

    def translate(self, text, target="zh"):
        return self._get("translate_text", text=text, target=target).get("translated", "")

    def models(self, tier=None):
        """全部可调模型 + 每百万 token 单价,按成本选型用。"""
        return self._get("list_models", tier=tier).get("models", [])

    def ask(self, prompt, model=None, max_tokens=None):
        """单模型问答。返回 dict:answer / cost_usd / latency_ms。"""
        return self._get("ask_model", prompt=prompt, model=model, max_tokens=max_tokens)

    def compare(self, prompt, models=None):
        """同一个提示词跑多个模型,返回各自答案 + 真实计量成本 + 延迟。"""
        return self._get("compare_models", prompt=prompt,
                         models=",".join(models) if models else None)

    # —— 写入/长任务类:走 MCP ——
    def remember(self, content, kind="note", project=None):
        """存一条会跨会话保留的记忆。带 Key 时跨设备、跨 agent 共享。"""
        return self._rpc("remember", {"content": content, "kind": kind, "project": project})

    def recall(self, query=None, project=None):
        return self._rpc("recall", {k: v for k, v in
                                    {"query": query, "project": project}.items() if v}).get("memories", [])

    def build(self, description, refine=None, visibility="public", wait=True, poll=10):
        """★ 一句话造一个真正在线的应用。约 100 秒后返回 app_url;wait=False 则立刻返回 job_id。"""
        job = self._rpc("build_app", {k: v for k, v in
                                      {"description": description, "refine": refine,
                                       "visibility": visibility}.items() if v})
        if not wait:
            return job
        return self.wait_job(job["job_id"], poll=poll)

    def wait_job(self, job_id, poll=10, timeout=600):
        """轮询长任务直到结束。"""
        deadline = time.time() + timeout
        while time.time() < deadline:
            view = self._get("check_job", job_id=job_id)
            if view.get("status") == "done":
                return view.get("structured_result") or view
            if view.get("status") == "error":
                raise NetCafeError(view.get("error", "job failed"), view)
            time.sleep(poll)
        raise NetCafeError(f"job {job_id} did not finish in {timeout}s")


if __name__ == "__main__":
    nc = NetCafe()
    print("可用模型:", [m["id"] for m in nc.models()][:5])
    print("图表 URL:", nc.diagram_url("graph TD; 想法-->应用"))

File: netcafe.mjs
/**
 * netcafe.mjs — 给 AI agent 用的一套工具,单文件、零依赖(只用内置 fetch)。
 *
 * 不用注册、不用密钥就能开始:匿名有免费额度,每次调用都会告诉你花了多少钱。
 * 额度不够时传一把 AllRouter Key 进来即可,调用方式完全不变。
 *
 *   import { NetCafe } from './netcafe.mjs';
 *   const nc = new NetCafe();                  // 匿名
 *   const nc = new NetCafe({ key: 'sk-...' }); // 带 Key:额度更大 + 跨设备记忆
 *
 *   await nc.search('MCP 协议 2026');
 *   await nc.read('https://example.com');
 *   nc.diagramUrl('graph TD; A-->B');          // 不发请求,直接拼出可嵌 Markdown 的图片 URL
 *   await nc.build('一个番茄钟,可自定义时长');   // ★ 一句话造一个真正在线的应用
 *
 * 全部工具:https://ainetcafe.com/mcp.html
 */
export const VERSION = '1.1.1';
const BASE = 'https://ainetcafe.com';

export class NetCafeError extends Error {
  constructor(message, payload = {}) {
    // message 里带上给人看的出路:撞墙后 agent 的默认行为是默默重试,
    // 结构化字段人看不到,异常打印的那一行是唯一保证到人眼前的位置。
    const extra = payload.tell_your_human;
    super(extra && !String(message).includes(extra) ? `${message}\n>>> ${extra}` : message);
    this.name = 'NetCafeError';
    this.payload = payload;
    this.pairUrl = payload.pair_url;
    this.howToContinue = payload.how_to_continue;
    this.tellYourHuman = payload.tell_your_human;
  }
}

export class NetCafe {
  constructor({ key = null, base = BASE, timeout = 120000, source = 'jssdk' } = {}) {
    this.key = key; this.base = base.replace(/\/$/, ''); this.timeout = timeout; this.source = source;
  }
  #headers(extra = {}) {
    return { 'user-agent': `netcafe-js/${VERSION}`, accept: 'application/json',
      ...(this.key ? { authorization: `Bearer ${this.key}` } : {}), ...extra };
  }
  async #get(tool, params = {}) {
    const q = new URLSearchParams();
    for (const [k, v] of Object.entries(params)) if (v !== undefined && v !== null) q.set(k, String(v));
    if (!q.has('s')) q.set('s', this.source);
    const r = await fetch(`${this.base}/t/${tool}?${q}`, { headers: this.#headers(),
      signal: AbortSignal.timeout(this.timeout) });
    const d = await r.json().catch(() => ({ error: `HTTP ${r.status}` }));
    if (!r.ok || d.error) throw new NetCafeError(d.error || `HTTP ${r.status}`, d);
    return d;
  }
  async #rpc(tool, args) {
    const r = await fetch(`${this.base}/mcp`, { method: 'POST',
      headers: this.#headers({ 'content-type': 'application/json', accept: 'application/json, text/event-stream' }),
      body: JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'tools/call', params: { name: tool, arguments: args } }),
      signal: AbortSignal.timeout(this.timeout) });
    const d = await r.json();
    const res = d.result || {};
    let payload = res.structuredContent;
    if (!payload) { try { payload = JSON.parse(res.content?.[0]?.text || '{}'); } catch { payload = {}; } }
    if (payload.error) throw new NetCafeError(payload.error, payload);
    return payload;
  }
  // —— 读取类 ——
  async search(query, maxResults = 8) { return (await this.#get('web_search', { query, max_results: maxResults })).results || []; }
  async read(url) { return (await this.#get('fetch_page', { url })).markdown || ''; }
  async diagram(source, type = 'mermaid', format = 'svg') { return (await this.#get('render_diagram', { source, type, format })).image_url; }
  /** 不发请求,直接拼出无状态图片 URL —— 适合写死在 README / 文档里 */
  diagramUrl(source, type = 'mermaid', format = 'svg') {
    const b64 = Buffer.from(source, 'utf8').toString('base64url');
    return `${this.base}/i/${type}/${b64}?format=${format}`;
  }
  async grammar(text, language = 'auto') { return (await this.#get('check_grammar', { text, language })).issues || []; }
  async transcribe(url, language) { return (await this.#get('transcribe_audio', { url, language })).text || ''; }
  async pdf({ url, html } = {}) { return (await this.#get('convert_to_pdf', { url, html })).pdf_url; }
  async translate(text, target = 'zh') { return (await this.#get('translate_text', { text, target })).translated || ''; }
  async models(tier) { return (await this.#get('list_models', { tier })).models || []; }
  /** 每个模型「一次调用」的真实美元成本(来自平台账单,非厂商标价)。CC BY 4.0。 */
  async costs(days = 30) { return (await this.#get('model_costs', { days })).models || []; }
  async ask(prompt, { model, maxTokens } = {}) { return this.#get('ask_model', { prompt, model, max_tokens: maxTokens }); }
  async compare(prompt, models) { return this.#get('compare_models', { prompt, models: models?.join(',') }); }
  // —— 写入/长任务 ——
  async remember(content, { kind = 'note', project } = {}) { return this.#rpc('remember', { content, kind, project }); }
  async recall(query, project) { return (await this.#rpc('recall', { ...(query ? { query } : {}), ...(project ? { project } : {}) })).memories || []; }
  /** ★ 一句话造一个真正在线的应用。约 100 秒;wait:false 则立刻返回 job_id */
  async build(description, { refine, visibility = 'public', wait = true, poll = 10000 } = {}) {
    const job = await this.#rpc('build_app', { description, ...(refine ? { refine } : {}), visibility });
    return wait ? this.waitJob(job.job_id, { poll }) : job;
  }
  async waitJob(jobId, { poll = 10000, timeout = 600000 } = {}) {
    const deadline = Date.now() + timeout;
    while (Date.now() < deadline) {
      const v = await this.#get('check_job', { job_id: jobId });
      if (v.status === 'done') return v.structured_result || v;
      if (v.status === 'error') throw new NetCafeError(v.error || 'job failed', v);
      await new Promise((r) => setTimeout(r, poll));
    }
    throw new NetCafeError(`job ${jobId} did not finish in time`);
  }
}
export default NetCafe;
